File: frontend/src/engine/sCurve.ts
/**
 * Client-side S-curve distribution for inline sparkline previews.
 * The Python backend engine is the source of truth for actual calculations.
 */

export function sCurveDistribution(
  totalAmount: number,
  durationMonths: number,
  steepness: number = 5
): number[] {
  if (durationMonths <= 0) return [];
  if (durationMonths === 1) return [totalAmount];

  const k = steepness;
  const midpoint = durationMonths / 2;

  // Cumulative S-curve values
  const cumulative: number[] = [];
  for (let m = 0; m <= durationMonths; m++) {
    cumulative.push(1 / (1 + Math.exp(-k * (m - midpoint) / durationMonths * 4)));
  }

  // Normalize to get monthly increments
  const totalCurve = cumulative[durationMonths] - cumulative[0];
  const monthly: number[] = [];
  for (let m = 0; m < durationMonths; m++) {
    monthly.push(((cumulative[m + 1] - cumulative[m]) / totalCurve) * totalAmount);
  }

  // Adjust rounding
  const diff = totalAmount - monthly.reduce((a, b) => a + b, 0);
  monthly[monthly.length - 1] += diff;

  return monthly;
}

export function straightLineDistribution(
  totalAmount: number,
  durationMonths: number
): number[] {
  if (durationMonths <= 0) return [];
  if (durationMonths === 1) return [totalAmount];

  const monthlyAmount = totalAmount / durationMonths;
  const monthly = Array(durationMonths).fill(monthlyAmount);

  const diff = totalAmount - monthly.reduce((a: number, b: number) => a + b, 0);
  monthly[monthly.length - 1] += diff;

  return monthly;
}

File: backend/app/engine/s_curve.py
import numpy as np
from typing import List


def s_curve_distribution(total_amount: float, duration_months: int, steepness: int = 5) -> List[float]:
    """
    Distribute total_amount across duration_months using logistic S-curve.
    steepness: 1 (flat/linear) to 9 (steep/concentrated in middle)
    Returns list of monthly amounts that sum to total_amount.
    """
    if duration_months <= 0:
        return []
    if duration_months == 1:
        return [total_amount]

    k = steepness
    midpoint = duration_months / 2
    months = range(duration_months)

    # Cumulative S-curve values
    cumulative = [
        1 / (1 + np.exp(-k * (m - midpoint) / duration_months * 4))
        for m in range(duration_months + 1)
    ]

    # Normalize to get monthly increments
    total_curve = cumulative[-1] - cumulative[0]
    monthly = [
        (cumulative[m + 1] - cumulative[m]) / total_curve * total_amount
        for m in months
    ]

    # Adjust rounding to match total exactly
    diff = total_amount - sum(monthly)
    monthly[-1] += diff

    return monthly


def straight_line_distribution(total_amount: float, duration_months: int) -> List[float]:
    """
    Distribute total_amount evenly across duration_months.
    Returns list of monthly amounts that sum to total_amount.
    """
    if duration_months <= 0:
        return []
    if duration_months == 1:
        return [total_amount]

    monthly_amount = total_amount / duration_months
    monthly = [monthly_amount] * duration_months

    # Adjust rounding
    diff = total_amount - sum(monthly)
    monthly[-1] += diff

    return monthly
